### PR TITLE
RSDK-2655 Change RPC server to be owned by resource servers

### PR DESCRIPTION
**Pull request by benjirewis**
[RSDK-2655](https://viam.atlassian.net/browse/RSDK-2655)

Main changes:
- Removes all `register_server` methods
- Changes constructors of all children of `ResourceServer` to take in a `shared_ptr<Server>`
- Stores that server as `server_` in the `ResourceServer`
- Registers the relevant gRPC service for a server in `ResourceServer` construction
- Changes `create_resource_server` methods to take in a `shared_ptr<Server>`
- Calls the modified `create_resource_server` method in `ModuleService::add_api_from_registry_inlock_`

Secondary changes:
- Removes almost all `server_to_mock_pipeline` functions in favor of a further abstracted `client_to_mock_pipeline` function added to `test_utils`
- Adds a constructor to the `MockRobotService` to be used in testing
- Removes some (seemingly) duplicative tests and mock code

[RSDK-2655]: https://viam.atlassian.net/browse/RSDK-2655?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

cc @zaporter-work for interest

### CODE REVIEW

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (benjirewis) [inline src/viam/examples/modules/complex/gizmo/api.cpp:27]: As mentioned in the PR description, general pattern here for all `create_resource_server` methods is to accept a `std::shared_ptr<Server>` and pass it to construction of the `[TypeOfResource]Server`.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (benjirewis) [inline src/viam/examples/modules/complex/gizmo/api.cpp:61]: As mentioned in the PR description, general pattern here for all `[TypeOfResource]Server` constructors is to invoke the `ResourceServer` constructor, pass along the resource manager, and server, and also pass the `[TypeOfResource]Server` itself with `this`. Once we're in `resource_server_base`, `this` seems to no longer qualify as a gRPC service and cannot be registered via `register_service()`, so I chose to just pass `this` as a `grpc::Service*`.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (benjirewis) [inline src/viam/examples/modules/complex/gizmo/api.cpp:199]: `register_server` is now useless, as `register_service` is called in the `ResourceServer` constructor.

File: src/viam/examples/modules/complex/proto/buf.lock
Comment (benjirewis) [inline src/viam/examples/modules/complex/proto/buf.lock:8]: Another googleapis buf bump it seems 🤷🏻 .

File: src/viam/examples/modules/complex/test_complex_module.cpp
Comment (benjirewis) [inline src/viam/examples/modules/complex/test_complex_module.cpp:75]: The new `client_to_mock_pipeline` requires specifying your `Client` and `Server` types and passing in your `mock` resource. The second argument lambda has access to the created client, and the mock can be accessed if needed by capturing by reference (`[&]`, I tried to use `[]` when not necessary)

File: src/viam/sdk/rpc/server.cpp
Comment (benjirewis) [inline src/viam/sdk/rpc/server.cpp:60]: `client_to_mock_pipeline` needs access to an in process channel, so I exposed one on the `sdk::Server::server_` with this method.

File: src/viam/sdk/tests/mocks/camera_mocks.cpp
Comment (benjirewis) [inline src/viam/sdk/tests/mocks/camera_mocks.cpp:116]: For both `camera_mocks` and `generic_mocks`, you'll see that I've removed mock client and server classes and their relevant code. From what I understand, a single test of a client method with `client_to_mock_pipeline` _should_ exercise client, server, and actual resource code, and I didn't think three separate tests (leveraging three classes like `MockCamera`, `MockCameraClient` and `MockCameraStub`) for each method was necessary. I could _totally_ be wrong in that assumption, and am happy to add this code back if desired.

File: src/viam/sdk/tests/mocks/generic_mocks.cpp
Comment (benjirewis) [inline src/viam/sdk/tests/mocks/generic_mocks.cpp:27]: Got a weird IDE error without this on the return of `generic` below: `Constness of 'generic' prevents automatic move (clang-tidy performance-no-automatic-move)` .

File: src/viam/sdk/tests/mocks/mock_robot.hpp
Comment (benjirewis) [inline src/viam/sdk/tests/mocks/mock_robot.hpp:21]: Needed this constructor for `robot_client_to_mocks_pipeline` and it had to be public.

File: src/viam/sdk/tests/test_robot.cpp
Comment (benjirewis) [inline src/viam/sdk/tests/test_robot.cpp:38]: Couldn't quite replace this one with `client_to_mocks_pipeline`  as it was a subtly different setup; see my comment above.

File: src/viam/sdk/tests/test_utils.hpp
Comment (benjirewis) [inline src/viam/sdk/tests/test_utils.hpp:39]: Stole this from @zaporter-work's pattern and made it more templated.

**Review by benjirewis — COMMENTED** [review]
😮‍💨 this one had a lot of "decisions", let me know if you find anything strange.

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [inline src/viam/sdk/module/service.cpp:275]: I _thought_ I could make the `server_` field of `ModuleService` a `unique_ptr`, but I think we have to share it due to this call (a single `Server` needs to be shared between between potentially multiple `[TypeOfResource]Server` types).

File: src/viam/sdk/tests/test_robot.cpp
Comment (stuqdog) [reply src/viam/sdk/tests/test_robot.cpp:38]: Is the issue just that we register multiple resources? We could make the `client_to_mocks_pipeline` take a vec of resources to register rather than a single resource and register everything in the vec... not sure this is actually better though, given the sort of unique space that the `robot` occupies and the slightly weird structure of having every other test file pass a vec with only a single element.

I think I'm convincing myself that it's fine as is, but including comment for posterity/to see if others have thoughts.

**Review by stuqdog — CHANGES_REQUESTED** [review]
Nice, looks good! Would like to see us not expose the `grpc_in_process_channel` method from `Server` and instead use a `friend struct` in tests, otherwise this is great.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (benjirewis) [reply src/viam/examples/modules/complex/gizmo/api.cpp:61]: Yeah that's a good idea. So I took a wack at this: it would certainly be a bit clearer, but it also does not work for some reason. gRPC service registration seems to break a bit and it's nontrivial for me to tell why. All I can see is that every test fails because any call to one our resource clients returns an exception (can't even introspect the error message for some reason). Everything _compiles_ fine but 🤷🏻 . If you feel strongly that we should try to adopt the pattern you're suggesting (which I do think is a good one), I could make a followup ticket for it?

File: src/viam/examples/dial/example_dial.cpp
Comment (benjirewis) [inline src/viam/examples/dial/example_dial.cpp:71]: Spoke with @stuqdog offline. Robot clients are closed in the types constructors, so this line leads to a double close and a "broken pipe" error; we should remove it.

File: src/viam/examples/modules/complex/gizmo/api.hpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/gizmo/api.hpp:68]: Various thoughts here:

- Feels boilerplate-ish if every `[TypeOfResource]Server` needs to declare two nearly identical constructors. What's the use-case for each? Is there a way we could combine them?

```
GizmoServer(std::shared_ptr<Server> server, std::shared_ptr<ResourceManager> manager = {}) : 
    ResourceServer(server, manager ? manager : std::make_shared<ResourceManager>()) {}
```

That still feels chatty. Could `ResourceServer` do the same trick instead?

- If there are going to be two signatures, I'd put the common parameter first (`std::shared_ptr<Server>`).

- I see a lot of places where `ResourceManager` and `Server` travel around together. Is there a missing abstraction? Why is there only sometimes a `ResourceManager`?

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/gizmo/api.cpp:26]: [nit]: `std::move(manager), std::move(server)`. You can argue that it makes no difference here and that'd be pretty true, but its a good habit to be consistent about using move. There's a lot of refcount jiggling happening here for basically no reason.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/gizmo/api.cpp:59]: `std::move(server)`, here and below both that and `std::move(manager)`.

Comment (acmorrow) [inline src/viam/examples/modules/complex/gizmo/api.cpp:59]: You could delegate to the two parameter `GizmoServer` constructor here.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (acmorrow) [reply src/viam/examples/modules/complex/gizmo/api.cpp:199]: But the point of the virtual `register_server` was exactly to work around the issue you called out above, that up in `ResourceBase` the `this` pointer is not useable.

File: src/viam/sdk/resource/resource_server_base.hpp
Comment (acmorrow) [inline src/viam/sdk/resource/resource_server_base.hpp:17]: - Please move the implementation out of line.
- `std::move` on `manager` and `server`.
- Have some doubts about this approach of having the `grpc::Service*` become part of the constructor signature for an API type.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (acmorrow) [reply src/viam/examples/modules/complex/gizmo/api.cpp:61]: I'm not sure it is a good idea to make our classes derive from gRPC classes that way. I think composition, as we are doing now, is going to get you further in the long run and better insulate you from changes in the gRPC world.

File: src/viam/examples/modules/complex/summation/api.cpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/summation/api.cpp:26]: Same comments here re moves, ctor signature, etc as in Gizmo stuff.

File: src/viam/examples/modules/complex/gizmo/api.hpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/gizmo/api.hpp:67]: I think this ctor is the more important one to make `explicit`.

File: src/viam/examples/modules/complex/summation/api.hpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/summation/api.hpp:63]: `explicit`

File: src/viam/examples/modules/complex/test_complex_module.cpp
Comment (acmorrow) [inline src/viam/examples/modules/complex/test_complex_module.cpp:74]: I think there is a nice trick you can use to save some typing:

Up where you had `gizmo_server_to_mock_pipeline` before, do:

```
template<typename Lambda>
using gizmo_server_to_mock_pipeline = client_to_mock_pipeline<GizmoClient, GizmoServer, Lambda>;
```

Then you don't need to re-iterate the template parameters for the client and server everywhere. Maybe mod a rename if you don't like the name `gizmo_server_to_mock_pipeline` and want to call it `gizmo_client_to_mock_pipeline` or something.

File: src/viam/examples/dial/example_dial.cpp
Comment (acmorrow) [reply src/viam/examples/dial/example_dial.cpp:71]: Does that mean that the `close` method on `RobotClient` is dangerous and should be removed? Or, should it be ok to `close` a `RobotClient`, and the dtor for `RobotClient` should deal with it more gracefully?

File: src/viam/sdk/module/service.cpp
Comment (acmorrow) [reply src/viam/sdk/module/service.cpp:275]: I think that depends on whether it is necessary for `[TypeOfResource]Server` objects to potentially extend the lifetime of the `Server` after the `ModuleService` has been destroyed. Does that actually make sense? Otherwise, the `ModuleService` could just own the `Server` object and you could pass it by bare pointer to `[TypeOfResource]Server` constructors. Their lifetime would need to be a subset of the `ModuleService`.

File: src/viam/sdk/rpc/server.hpp
Comment (acmorrow) [inline src/viam/sdk/rpc/server.hpp:6]: Why?

File: src/viam/sdk/tests/mocks/generic_mocks.cpp
Comment (acmorrow) [reply src/viam/sdk/tests/mocks/generic_mocks.cpp:27]: It isn't that weird. It is telling you that it will need to return by copy rather than move, since you can't move from `const`. That's a slight performance penalty, so it is warning you. I think it is a good warning and hopefully clang-tidy will tell us about it.

File: src/viam/sdk/tests/test_robot.cpp
Comment (acmorrow) [reply src/viam/sdk/tests/test_robot.cpp:39]: I think it is fine how you have it.

File: src/viam/sdk/tests/test_utils.hpp
Comment (acmorrow) [inline src/viam/sdk/tests/test_utils.hpp:56]: `auto`'d be fine here.

File: src/viam/sdk/tests/test_utils.hpp
Comment (acmorrow) [inline src/viam/sdk/tests/test_utils.hpp:64]: `auto` would be fine here. Also, did you intend for the lifetime of `TestServer` to end on this line? I think it works because you extract a shared pointer. But if you ever wanted `TestServer` to do more (like hold some state) I think this would be fragile. Maybe just make a named `TestServer` object on the stack?

File: src/viam/examples/dial/example_dial.cpp
Comment (benjirewis) [reply src/viam/examples/dial/example_dial.cpp:71]: It's a good question; @stuqdog would know more about the semantics of `close()`. Judging from [the doc comments](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/robot/client.hpp#L42-L44), the necessity of calling `close` seems to depend on which way you've constructed the `RobotClient`.

File: src/viam/examples/modules/complex/gizmo/api.cpp
Comment (benjirewis) [reply src/viam/examples/modules/complex/gizmo/api.cpp:199]: I'd agree that `register_server()` used to be the way that you could register a service for a `Server`; my "replacement" for this logic is to pass `this` up to `ResourceServerBase` (which is also not ideal). I don't mean to say this was a bad design or anything, just that `register_service()` gets called in `ResourceServerBase` construction now. Did you have a reason to keep these methods around?

File: src/viam/examples/modules/complex/gizmo/api.hpp
Comment (benjirewis) [reply src/viam/examples/modules/complex/gizmo/api.hpp:68]: Yeah now that you mention it, I see no usages of the `[TypeOfResource]Server(std::shared_ptr<Server> server)` constructors. In other words, I think you are correct that we always create resource servers with both a `ResourceManager` and `Server`. I _could_ remove all the `[TypeOfResource]Server(std::shared_ptr<Server> server)` constructors, but that would be hypothetically backward breaking. In any case, I'll try an approach like what you've suggested with default values for `manager` and a ternary operator to make one if none is passed in (hopefully at the `ResourceServerBase` level to be as simple as possible).

File: src/viam/sdk/module/service.cpp
Comment (benjirewis) [reply src/viam/sdk/module/service.cpp:275]: > Does that actually make sense? 

I don't think `[TypeOfResource]Server`s should ever outlive the `ModuleService` so that's a good point. I'll try passing by bare pointer.

File: src/viam/sdk/resource/resource_server_base.hpp
Comment (benjirewis) [reply src/viam/sdk/resource/resource_server_base.hpp:17]: Can do the first two.

> Have some doubts about this approach of having the grpc::Service* become part of the constructor signature for an API type.

Did you have a different idea here?

File: src/viam/sdk/tests/test_utils.hpp
Comment (benjirewis) [reply src/viam/sdk/tests/test_utils.hpp:64]: > Also, did you intend for the lifetime of TestServer to end on this line?

I did, but I can make an actual object on the stack instead.

**Review by benjirewis — COMMENTED** [review]
Thanks for comments @acmorrow , still working through them but had a couple responses/qs.

File: src/viam/examples/modules/complex/gizmo/api.hpp
Comment (benjirewis) [reply src/viam/examples/modules/complex/gizmo/api.hpp:68]: Actually I'll hold off on this until @stuqdog weighs in on whether we can remove `[TypeOfResource]Server` ctors that take no resource manager.

File: src/viam/examples/modules/complex/test_complex_module.cpp
Comment (benjirewis) [reply src/viam/examples/modules/complex/test_complex_module.cpp:74]: Spoke a bit offline; this would be cool, but it doesn't fully work as written there. `client_to_mock_pipeline` is a function not a type, so it can't be on the right hand side of a `using` statement. I may just leave the unfortunately verbose `client_to_mock_pipeline<*Client, *Server>` pattern around.

File: src/viam/sdk/rpc/server.hpp
Comment (benjirewis) [reply src/viam/sdk/rpc/server.hpp:6]: Not needed; good catch!

File: src/viam/sdk/tests/test_utils.hpp
Comment (benjirewis) [reply src/viam/sdk/tests/test_utils.hpp:56]: Done.

File: src/viam/examples/dial/example_dial.cpp
Comment (stuqdog) [reply src/viam/examples/dial/example_dial.cpp:71]: That's correct. The thinking is/was, if people are asking us to build a channel for them (via the `at_address` call) then it makes sense for us to `close` that channel when the client is done with it. If a user is providing their own channel (via the `with_channel` call) then we don't want to make assumptions about that when to close that channel.

That said, I do think we may be painting with too broad a brush here. Just because the channel isn't being closed doesn't necessarily mean we shouldn't call the other aspects of `close` (namely, the `~thread()` calls). So I do think there's room for making `close` cleaner/making the `RobotClient` dtor to be more graceful in its handling of it.

File: src/viam/examples/modules/complex/gizmo/api.hpp
Comment (stuqdog) [reply src/viam/examples/modules/complex/gizmo/api.hpp:68]: > whether we can remove [TypeOfResource]Server ctors that take no resource manager.
I think this is fine. I agree we don't seem to be using the `manager`-less version anywhere, it's easy enough to make a `ResourceManager` if we need to, and the current structure is a bit noisy.

**Review by stuqdog — APPROVED** [review]
Interested in seeing how ongoing conversations (`register_server()`, etc.) get resolved, but otherwise lgtm.

File: src/viam/examples/dial/example_dial.cpp
Comment (benjirewis) [reply src/viam/examples/dial/example_dial.cpp:71]: Gotcha; thanks for the context! Filed https://viam.atlassian.net/browse/RSDK-6212, as it seems slightly out of scope for this PR.

**Comment by benjirewis** [comment]
Closing in favor of #199 .

**Pull request closed without merging**
